backends/tfhe-cuda-backend/cuda/src/crypto/keyswitch.cuh

```diff
@@ -146,12 +147,169 @@ void execute_keyswitch_async(cudaStream_t *streams, uint32_t *gpu_indexes,
         GET_VARIANT_ELEMENT(lwe_input_indexes, i);
 
     // Compute Keyswitch
-    cuda_keyswitch_lwe_ciphertext_vector<Torus>(
+    host_keyswitch_lwe_ciphertext_vector<Torus>(
         streams[i], gpu_indexes[i], current_lwe_array_out,
         current_lwe_output_indexes, current_lwe_array_in,
         current_lwe_input_indexes, ksks[i], lwe_dimension_in, lwe_dimension_out,
         base_log, level_count, num_samples_on_gpu);
   }
 }
 
+// chunk_count = glwe_size * polynomial_size / threads.
+// each threads will responsible to process only lwe_size times multiplication
+template <typename Torus>
+__device__ void keyswitch_lwe_ciphertext_into_glwe_ciphertext(
+    Torus *glwe_out, Torus *lwe_in, Torus *fp_ksk, uint32_t lwe_dimension_in,
+    uint32_t glwe_dimension, uint32_t polynomial_size, uint32_t base_log,
+    uint32_t level_count) {
+
+  extern __shared__ int8_t sharedmem[];
+
+  // result accumulator, shared memory is used because of frequent access
+  Torus *local_glwe_chunk = (Torus *)sharedmem;
```

Comment:
I'm not sure this is actually more efficient than directly accessing global memory, have you benchmarked it? Since the data has to be copied into shared memory and back to global memory, it may just make no difference.

Reply:
I think I will merge as it is and try to remove shared memory on a next PR. I opened [an issue](https://github.com/zama-ai/tfhe-rs-internal/issues/678) about that.

Reply:
I will explore this in https://github.com/zama-ai/tfhe-rs-internal/issues/678.